<a href="https://colab.research.google.com/github/ChristianBridge/Christian_INFO5731_SPRING2025/blob/main/Bridge_Christian_Assignment_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment: 4**

**This exercise will provide a valuable learning experience in working with text data and extracting features using various topic modeling algorithms. Key concepts such as Latent Dirichlet Allocation (LDA), Latent Semantic Analysis (LSA) and BERTopic.**



**Expectations**:
*   Students are expected to complete the exercise during lecture period to meet the active participation criteria of the course.
*   Use the provided .*ipynb* document to write your code & respond to the questions. Avoid generating a new file.
*   Write complete answers and run all the cells before submission.
*   Make sure the submission is "clean"; *i.e.*, no unnecessary code cells.
*   Once finished, allow shared rights from top right corner (*see Canvas for details*).

**Total points**: 100


NOTE: The output should be presented well to get **full points**

**Late submissions will have a penalty of 10% of the marks for each day of late submission, and no requests will be answered. Manage your time accordingly.**


## Troubleshooting Numpy & Gensim versions

In [6]:
#Check current version of python, 3.11.x is okay
!python --version

Python 3.11.11


In [7]:
#The following libraries appear to be tethered to a version of numpy greater
#than 1.25, however we need 1.25, so we will uninstall these
#'''
!pip uninstall treescope -y
!pip uninstall tensorflow -y
!pip uninstall blosc2 -y
!pip uninstall dopamine-rl -y
!pip uninstall flax -y
!pip uninstall tables -y
#'''

Found existing installation: treescope 0.1.9
Uninstalling treescope-0.1.9:
  Successfully uninstalled treescope-0.1.9
Found existing installation: tensorflow 2.18.0
Uninstalling tensorflow-2.18.0:
  Successfully uninstalled tensorflow-2.18.0
Found existing installation: blosc2 3.2.1
Uninstalling blosc2-3.2.1:
  Successfully uninstalled blosc2-3.2.1
Found existing installation: dopamine_rl 4.1.2
Uninstalling dopamine_rl-4.1.2:
  Successfully uninstalled dopamine_rl-4.1.2
Found existing installation: flax 0.10.5
Uninstalling flax-0.10.5:
  Successfully uninstalled flax-0.10.5
Found existing installation: tables 3.10.2
Uninstalling tables-3.10.2:
  Successfully uninstalled tables-3.10.2


In [1]:
#Force our numpy version to be 1.25 which is compatible with gensim
!pip install --force-reinstall numpy==1.25.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.6/17.6 MB 37.0 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 2.0.2
    Uninstalling numpy-2.0.2:
      Successfully uninstalled numpy-2.0.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
treescope 0.1.9 requires numpy>=1.25.2, but you have numpy 1.25.0 which is incompatible.
tensorflow 2.18.0 requires numpy<2.1.0,>=1.26.0, but you have numpy 1.25.0 which is incompatible.
blosc2 3.2.1 requires numpy>=1.26, but you have numpy 1.25.0 which is incompatible.


In [1]:
#Install gensim after numpy
!pip install gensim -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.7/26.7 MB 31.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 38.6/38.6 MB 11.4 MB/s eta 0:00:00


In [2]:
#Install LDA Visualizer
!pip install pyLDAvis -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 15.8 MB/s eta 0:00:00


In [41]:
!pip install bertopic -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.6/150.6 kB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 54.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 34.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 44.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 846.0 kB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 48.9 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstal

In [3]:
#Used for OpenAI Integration
!pip install tiktoken openai -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 11.5 MB/s eta 0:00:00


## Import Libraries

In [1]:
#Check installed versions, Numpy 1.25.x is compatible with gensim 4.3.3
import numpy
import gensim
print("NumPy version:", numpy.__version__)
print("Gensim version:", gensim.__version__)

NumPy version: 1.25.0
Gensim version: 4.3.3


In [19]:
#Import Libraries
import re
import random
import pandas as pd
from sklearn.datasets import fetch_20newsgroups
from gensim import corpora
from gensim.models import LsiModel
from gensim.models.coherencemodel import CoherenceModel
from bertopic import BERTopic
from gensim.utils import simple_preprocess
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
from pprint import pprint
import pyLDAvis
import pyLDAvis.gensim_models as gensimvis

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


# **Question 1 (20 Points)**

**Dataset**: 20 Newsgroups dataset

**Dataset Link**: https://scikit-learn.org/0.19/datasets/twenty_newsgroups.html

**Consider Random 2000 rows only**

Generate K=10 topics by using LDA and LSA,
then calculate coherence score and determine the optimized K value by the coherence score. Further, summarize and visualize each topics in you own words.


## Select Random 2000 rows for dataset

In [3]:
newsgroups = fetch_20newsgroups(subset = 'train')

In [4]:
newsgroups.filenames.shape

(11314,)

In [5]:
def random_sample(dataset):
    random_list = []
    random_data = []

    num = 2000
    shape = dataset.filenames.shape[0] - 1
    num_samples = 0

    while num_samples < num:
        rand_num = random.randint(0,shape)
        if rand_num in random_list:
            continue
        else:
            random_list.append(rand_num)
            random_data.append(dataset.data[rand_num])
            num_samples = len(random_list)

    return random_list, random_data

In [6]:
data_list, news_data = random_sample(newsgroups)

In [7]:
len(news_data)

2000

In [8]:
df = pd.DataFrame(news_data, columns = ['text'])

In [9]:
df['text']

0       From: jca2@cec1.wustl.edu (Joseph Charles Achk...
1       From: moseley@u.washington.edu (Steve L. Mosel...
2       From: geb@cs.pitt.edu (Gordon Banks)\nSubject:...
3       From: kelleyb@austin.ibm.com (Kelley Boylan)\n...
4       From: cramer@optilink.COM (Clayton Cramer)\nSu...
                              ...                        
1995    From: gt6511a@prism.gatech.EDU (COCHRANE,JAMES...
1996    From: prb@access.digex.com (Pat)\nSubject: Con...
1997    From: jason@ab20.larc.nasa.gov (Jason Austin)\...
1998    From: paladin@world.std.com (Thomas G Schlatte...
1999    From: "William K. Willis" <ww1a+@andrew.cmu.ed...
Name: text, Length: 2000, dtype: object

## Data Preprocessing

In [62]:
from gensim.parsing.preprocessing import remove_stopwords, strip_punctuation \
                                        , preprocess_string, strip_short, stem_text

# preprocess given text
def preprocess(text):

    # clean text based on given filters
    CUSTOM_FILTERS = [lambda x: x.lower(),
                                remove_stopwords,
                                strip_punctuation,
                                strip_short,
                                stem_text]
    text = preprocess_string(text, CUSTOM_FILTERS)

    return text

# apply function to all reviews
df['cleaned_text'] = df['text'].apply(lambda x: preprocess(x))

In [11]:
df.head(5)

text  \
0  From: jca2@cec1.wustl.edu (Joseph Charles Achk...   
1  From: moseley@u.washington.edu (Steve L. Mosel...   
2  From: geb@cs.pitt.edu (Gordon Banks)\nSubject:...   
3  From: kelleyb@austin.ibm.com (Kelley Boylan)\n...   
4  From: cramer@optilink.COM (Clayton Cramer)\nSu...   

                                        cleaned_text  
0  [from, jca2, cec1, wustl, edu, joseph, charl, ...  
1  [from, moselei, washington, edu, steve, mosele...  
2  [from, geb, pitt, edu, gordon, bank, subject, ...  
3  [from, kelleyb, austin, ibm, com, kellei, boyl...  
4  [from, cramer, optilink, com, clayton, cramer,...

## Latent Semantic Analysis (LSA)

In [12]:
# Coherence score in topic modeling to measure how interpretable the topics are to humans.
# find the coherence score with a different number of topics

#Store the best model for future comparison/usage

#data = df['cleaned_text']
def lsa_model_k_finder(data, max_topics):
    # create a dictionary with the corpus
    corpus = data
    dictionary = corpora.Dictionary(corpus)

    # convert corpus into a bag of words
    bow = [dictionary.doc2bow(text) for text in corpus]

    #Initiate placeholder values for best model return
    lsi = LsiModel(bow, num_topics=1, id2word=dictionary)
    best_model = CoherenceModel(model = lsi, texts = data, dictionary = dictionary, coherence='c_v')
    best_coh_score = best_model.get_coherence()
    num_topics = 2

    #Iterate through K (topics)
    for i in range(2,(max_topics + 1)):
        lsi = LsiModel(bow, num_topics=i, id2word=dictionary)
        coherence_model = CoherenceModel(model = lsi, texts = data, dictionary = dictionary, coherence='c_v')
        coherence_score = coherence_model.get_coherence()
        print(f"Coherence score with {i} clusters: {coherence_score}")
        if coherence_score > best_coh_score:
            best_coh_score = coherence_score
            best_model = CoherenceModel(model = lsi, texts = data, dictionary = dictionary, coherence='c_v')
            num_topics = i
            print(f"Model with {i} topics and {coherence_score} is currently the best.")
        else:
            continue

    print(f"Returning Model with {best_coh_score} coherence score given {num_topics} topics.")

    return best_model, best_coh_score

In [13]:
lsa_model, lsa_score = lsa_model_k_finder(df['cleaned_text'],10)

Coherence score with 2 clusters: 0.71263927355711
Coherence score with 3 clusters: 0.6718766409766381
Coherence score with 4 clusters: 0.5431805626087168
Coherence score with 5 clusters: 0.6016064986319238
Coherence score with 6 clusters: 0.6149346855179965
Coherence score with 7 clusters: 0.537367080277414
Coherence score with 8 clusters: 0.6705683850103464
Coherence score with 9 clusters: 0.624852892946269
Coherence score with 10 clusters: 0.5975935588882545
Returning Model with 0.9262328602282258 coherence score given 2 topics.


##LDA

In [15]:
#Import stop words, and extend using common found e-mail terms
stop_words = stopwords.words('english')
stop_words.extend(['from', 'subject', 're', 'edu', 'use'])

#Changes sentences to words
def sent_to_words(sentences):
    for sentence in sentences:
        # deacc=True removes punctuations
        yield(gensim.utils.simple_preprocess(str(sentence), deacc=True))

#Removes stopwords from word list
def remove_stopwords(texts):
    return [[word for word in simple_preprocess(str(doc))
             if word not in stop_words] for doc in texts]

#Extract and clean data
data = df['cleaned_text'].values.tolist()
data_words = list(sent_to_words(data))
# remove stop words
data_words = remove_stopwords(data_words)
print(data_words[:1][0][:30])

['jca', 'cec', 'wustl', 'joseph', 'charl', 'achkar', 'blue', 'steal', 'game', 'hawk', 'keyword', 'blue', 'hull', 'shanahan', 'joseph', 'blackhawk', 'belfour', 'nntp', 'post', 'host', 'cec', 'organ', 'washington', 'univers', 'loui', 'line', 'blue', 'score', 'power', 'plai']


In [17]:
# Create Dictionary
id2word = corpora.Dictionary(data_words)
# Create Corpus
texts = data_words
# Term Document Frequency
corpus = [id2word.doc2bow(text) for text in texts]

In [18]:
# number of topics
num_topics = 10
# Build LDA model
lda_model = gensim.models.LdaMulticore(corpus=corpus,
                                       id2word=id2word,
                                       num_topics=num_topics)
# Print the Keyword in the 10 topics
pprint(lda_model.print_topics())
doc_lda = lda_model[corpus]

[(0,
  '0.005*"line" + 0.005*"organ" + 0.005*"com" + 0.005*"time" + 0.004*"write" + '
  '0.004*"new" + 0.004*"think" + 0.004*"peopl" + 0.003*"know" + '
  '0.003*"articl"'),
 (1,
  '0.052*"max" + 0.006*"ei" + 0.006*"line" + 0.005*"peopl" + 0.004*"organ" + '
  '0.003*"com" + 0.003*"like" + 0.003*"univers" + 0.003*"post" + 0.003*"us"'),
 (2,
  '0.008*"organ" + 0.008*"line" + 0.005*"post" + 0.005*"articl" + '
  '0.005*"write" + 0.005*"com" + 0.004*"peopl" + 0.004*"univers" + '
  '0.004*"like" + 0.003*"apr"'),
 (3,
  '0.007*"line" + 0.006*"com" + 0.005*"univers" + 0.004*"organ" + 0.004*"like" '
  '+ 0.004*"write" + 0.004*"articl" + 0.004*"time" + 0.004*"peopl" + '
  '0.003*"new"'),
 (4,
  '0.010*"com" + 0.008*"organ" + 0.008*"line" + 0.006*"write" + 0.006*"post" + '
  '0.005*"articl" + 0.004*"host" + 0.004*"peopl" + 0.004*"univers" + '
  '0.004*"nntp"'),
 (5,
  '0.010*"com" + 0.007*"line" + 0.006*"organ" + 0.006*"write" + 0.005*"like" + '
  '0.005*"work" + 0.005*"articl" + 0.004*"peopl" + 0

In [20]:
#import pyLDAvis.gensim

pyLDAvis.enable_notebook()

LDAvis_prepared = pyLDAvis.gensim_models.prepare(lda_model, corpus, id2word)

pyLDAvis.display(LDAvis_prepared)

# **BERTopic**

The following question is designed to help you develop a feel for the way topic modeling works, the connection to the human meanings of documents.

Dataset from **assignment-3** (text dataset) .

> Dont use any custom datasets.


> Dataset must have 1000+ rows, no duplicates and null values



# **Question 2 (20 Points)**



Q2) **Generate K=10 topics by using BERTopic and then find optimal K value by the coherence score. Interpret each topic and visualize with suitable style.**

In [21]:
semantic_df = pd.read_csv("semantic_scholar_10000.csv")

In [22]:
semantic_df.dropna(axis = 0, inplace = True)

In [23]:
semantic_df.reset_index()

index                                   paperId  \
0         0  00000c33779acab142af6c7a6dae8b36fac0805d   
1         1  0000238f07f151172cf2602588ba762b55c8464b   
2         3  0000315635be19f6278dbc72597b3065fac405f0   
3         4  00005d68c6c7eb4d3c27da8242a30b9a498f991e   
4         5  00005f1b7e976068ca4b5a1b546d9945158b3bfc   
...     ...                                       ...   
7865   9991  032afbb7d1b11d069627b0b2ddcb53dadc3441ba   
7866   9993  032b14ee53946228b132f2e3d9789edef40413f6   
7867   9994  032b1d12a2d8a669a5ea97f30edd5c41d2cbe5df   
7868   9996  032b3905a76c716fba857c7878209840b9e46ada   
7869   9998  032b423054d52846f9306f23bdd07eccee35a5c8   

                                                    url  \
0     https://www.semanticscholar.org/paper/00000c33...   
1     https://www.semanticscholar.org/paper/0000238f...   
2     https://www.semanticscholar.org/paper/00003156...   
3     https://www.semanticscholar.org/paper/00005d68...   
4     https://www.semanticscholar.org/paper/00005f1b...   
...                                                 ...   
7865  https://www.semanticscholar.org/paper/032afbb7...   
7866  https://www.semanticscholar.org/paper/032b14ee...   
7867  https://www.semanticscholar.org/paper/032b1d12...   
7868  https://www.semanticscholar.org/paper/032b3905...   
7869  https://www.semanticscholar.org/paper/032b4230...   

                                               abstract  \
0     In the era of burgeoning electric vehicle (EV)...   
1     Background Meditation apps have surged in popu...   
2     Background Humans must be able to cope with th...   
3     The growing number of cloud-based services has...   
4     Diffusion generative models, a class of machin...   
...                                                 ...   
7865  Backgrounds Freezing of gait (FoG) is a common...   
7866  Speaker recognition depends on identifying the...   
7867  The latest technologies and communication prot...   
7868  In the context of increasing negative environm...   
7869  Some of the simplest tasks for a human to acco...   

                                                authors  \
0     ['Ahmad Almaghrebi', 'Kevin James', 'Fares al ...   
1     ['Christian A. Webb', 'M. Hirshberg', 'R. Davi...   
2     ['Nida Shafiq', 'Isma Hamid', 'Muhammad Asif',...   
3                  ['Iehab Alrassan', 'Asma Alqahtani']   
4     ['Lasse Kreimendahl', 'Mikhail Karnaukh', 'M. ...   
...                                                 ...   
7865  ['Jae-Min Park', 'Chang-Won Moon', 'Byung Chan...   
7866         ['Venkata Subba Reddy Gade', 'M. Sumathi']   
7867    ['Bilal Muhammad Khan', 'Muhammad Bilal Kadri']   
7868  ['Son Vu Hong Pham', 'Khoi Van Tien Nguyen', '...   
7869  ['Eshani Akanksha Bisht', 'Purnendu Prabhat', ...   

                                     processed abstract  
0     ['era', 'burgeon', 'electr', 'vehicl', 'ev', '...  
1     ['background', 'medit', 'app', 'surg', 'popula...  
2     ['background', 'human', 'must', 'abl', 'cope',...  
3     ['grow', 'number', 'cloudbas', 'servic', 'led'...  
4     ['diffus', 'gener', 'model', 'class', 'machin'...  
...                                                 ...  
7865  ['background', 'freez', 'gait', 'fog', 'common...  
7866  ['speaker', 'recognit', 'depend', 'identifi', ...  
7867  ['latest', 'technolog', 'commun', 'protocol', ...  
7868  ['context', 'increas', 'neg', 'environment', '...  
7869  ['simplest', 'task', 'human', 'accomplish', 'd...  

[7870 rows x 6 columns]

In [24]:
#Text stored as string in CSV, convert back to list
def back_to_list(text):
    return re.sub(r"['\[\]]", "", text)

semantic_df['processed abstract'] = semantic_df['processed abstract'].apply(back_to_list)

In [25]:
semantic_df = semantic_df.sample(2000)

In [132]:
#abstracts = semantic_df['processed abstract'].to_string()

In [26]:
topic_model = BERTopic(language="english")
topics, probs = topic_model.fit_transform(semantic_df['processed abstract'])

In [27]:
topic_model.get_topic_info()

Topic  Count                                   Name  \
0      -1    867                -1_learn_data_model_use   
1       0    435            0_patient_predict_use_model   
2       1     62          1_attack_detect_secur_network   
3       2     52       2_sentiment_social_opinion_tweet   
4       3     47           3_text_languag_word_translat   
5       4     37      4_optim_algorithm_problem_converg   
6       5     36         5_problem_algorithm_learn_game   
7       6     34          6_model_predict_random_forest   
8       7     30                7_cnn_imag_deep_network   
9       8     29     8_quantum_molecular_topolog_comput   
10      9     28           9_model_paramet_vector_learn   
11     10     27        10_channel_signal_radio_antenna   
12     11     26        11_crop_growth_plant_agricultur   
13     12     25              12_season_co_water_pollut   
14     13     25          13_stock_market_price_financi   
15     14     24            14_power_pv_forecast_energi   
16     15     24           15_speech_languag_sign_music   
17     16     23     16_materi_thermal_properti_conduct   
18     17     22             17_fl_feder_client_privaci   
19     18     22    18_design_optim_circuit_architectur   
20     19     18           19_imag_segment_region_color   
21     20     17  20_neural_riemannian_manifold_network   
22     21     17          21_student_educ_teacher_cours   
23     22     15            22_task_learn_model_languag   
24     23     12         23_featur_select_page_classifi   
25     24     12           24_fraud_card_credit_banknot   
26     25     12            25_ai_erp_profession_integr   
27     26     11        26_commun_attack_implement_data   
28     27     11        27_vehicl_detect_yolov5_monitor   

                                       Representation  \
0   [learn, data, model, use, machin, system, meth...   
1   [patient, predict, use, model, diseas, studi, ...   
2   [attack, detect, secur, network, threat, iot, ...   
3   [sentiment, social, opinion, tweet, news, twit...   
4   [text, languag, word, translat, approach, port...   
5   [optim, algorithm, problem, converg, bound, st...   
6   [problem, algorithm, learn, game, solv, solut,...   
7   [model, predict, random, forest, classif, scre...   
8   [cnn, imag, deep, network, convolut, neural, o...   
9   [quantum, molecular, topolog, comput, simul, p...   
10  [model, paramet, vector, learn, set, space, st...   
11  [channel, signal, radio, antenna, wifi, wirele...   
12  [crop, growth, plant, agricultur, stage, use, ...   
13  [season, co, water, pollut, model, wind, ocean...   
14  [stock, market, price, financi, predict, retai...   
15  [power, pv, forecast, energi, fault, renew, tr...   
16  [speech, languag, sign, music, speaker, system...   
17  [materi, thermal, properti, conduct, mechan, s...   
18  [fl, feder, client, privaci, updat, commun, tr...   
19  [design, optim, circuit, architectur, hardwar,...   
20  [imag, segment, region, color, method, detect,...   
21  [neural, riemannian, manifold, network, field,...   
22  [student, educ, teacher, cours, learn, teach, ...   
23  [task, learn, model, languag, convers, onlin, ...   
24  [featur, select, page, classifi, coresearch, t...   
25  [fraud, card, credit, banknot, financi, detect...   
26  [ai, erp, profession, integr, skill, arbitr, s...   
27  [commun, attack, implement, data, log, workloa...   
28  [vehicl, detect, yolov5, monitor, autonom, dev...   

                                  Representative_Docs  
0   [problem, despit, rather, success, applic, art...  
1   [abstract, object, provid, focus, evalu, predi...  
2   [rais, threat, distribut, denial, servic, ddo,...  
3   [sentiment, analysi, process, analyz, peopl, s...  
4   [recent, grow, interest, short, text, classif,...  
5   [largescal, nonconvex, nonsmooth, problem, att...  
6   [studi, discus, implement, decis, tree, algori...  
7   [variou, featur, variabl, screen, approach, pr...  
8   [evid, paradig

In [28]:
topic_model.visualize_topics()

In [29]:
topic_model.visualize_barchart(top_n_topics=12, n_words = 10, width = 350, height = 350)

In [30]:
topic_model.visualize_hierarchy(top_n_topics=12, width = 700, height = 700)

In [33]:
#Extract tokenized texts
texts = semantic_df['processed abstract'].apply(lambda x: x.split()).tolist()

#Extract the top words per topic from BERTopic
topics = topic_model.get_topics()
top_words_per_topic = [
    [word for word, _ in topic_model.get_topics()[topic_id]]
    for topic_id in topic_model.get_topics().keys()
    if topic_id != -1
]

#Create dictionary
dictionary = corpora.Dictionary(texts)

#Compute Coherence Score
coherence_model = CoherenceModel(
    topics=top_words_per_topic,
    texts=texts,
    dictionary=dictionary,
    coherence='c_v'
)

coherence_score = coherence_model.get_coherence()
print(f"Coherence Score: {coherence_score:.4f}")

Coherence Score: 0.2547


In [42]:
#data = semantic_df['processed abstract']
def BERTopic_coherence(data,num_topics):

    best_model = BERTopic(language="english")
    best_score = 0
    best_topic_num = 0
    best_topics = []
    best_probs = []

    for i in range(2, num_topics):
        topic_model = BERTopic(language="english", nr_topics = i)
        topics, probs = topic_model.fit_transform(data)

        #Extract tokenized texts
        texts = data.apply(lambda x: x.split()).tolist()

        #Extract the top words per topic from BERTopic
        topics = topic_model.get_topics()
        top_words_per_topic = [
            [word for word, _ in topic_model.get_topics()[topic_id]]
            for topic_id in topic_model.get_topics().keys()
            if topic_id != -1
        ]

        #Create dictionary
        dictionary = corpora.Dictionary(texts)

        #Compute Coherence Score
        coherence_model = CoherenceModel(
            topics=top_words_per_topic,
            texts=texts,
            dictionary=dictionary,
            coherence='c_v'
        )
        coherence_score = coherence_model.get_coherence()

        print(f"Model with {i} topics, coherence score is {coherence_score:}")

        if coherence_score > best_score:
            best_topic_num = i
            best_model = topic_model
            best_score = coherence_score
            best_topics = topics
            best_probs = probs
        else:
            continue

    print(f"Model with {best_topic_num} topics with coherence score {best_score}")
    return best_model, best_score, best_topic_num, best_topics, best_probs

In [43]:
BERTopic_coherence(semantic_df['processed abstract'],10)

Model with 2 topics, coherence score is 0.5046858475621662
Model with 3 topics, coherence score is 0.3811601194838713
Model with 4 topics, coherence score is 0.3924180719817278
Model with 5 topics, coherence score is 0.39972663194490055
Model with 6 topics, coherence score is 0.3206102320286227
Model with 7 topics, coherence score is 0.28144394674625145
Model with 8 topics, coherence score is 0.3797396125440558
Model with 9 topics, coherence score is 0.33081392899531975
Model with 2 topics with coherence score 0.5046858475621662


(<bertopic._bertopic.BERTopic at 0x7e8701d73390>,
 0.5046858475621662,
 2,
 {-1: [('learn', 0.05488967507294258),
   ('use', 0.04849675248303692),
   ('model', 0.046821508805114986),
   ('data', 0.046028995632614485),
   ('machin', 0.04411971968538582),
   ('method', 0.03309358139639312),
   ('algorithm', 0.029790081296134054),
   ('system', 0.02966258075009669),
   ('predict', 0.026679038529236136),
   ('perform', 0.0262897618636391)],
  0: [('model', 0.04960145210542409),
   ('use', 0.04957107669097967),
   ('learn', 0.04798850314933193),
   ('machin', 0.04013376549308432),
   ('data', 0.03535920386532845),
   ('predict', 0.031135509162081153),
   ('method', 0.029409858657700867),
   ('studi', 0.024689835502172807),
   ('algorithm', 0.023625575068903632),
   ('perform', 0.023410085440460162)]},
 array([0.54672312, 0.        , 0.8700289 , ..., 0.        , 0.        ,
        0.        ]))

# **Question 3 (25 points)**


**Dataset Link**: 20 Newsgroup Dataset (Random 2000 values)

Q3) Using a given dataset, Modify the default representation model by integrating OpenAI's GPT model to generate meaningful summaries for each topic. Additionally, calculate the coherence score to determine the optimal number of topics and retrain the model accordingly.



Usefull Link: https://maartengr.github.io/BERTopic/getting_started/representation/llm#truncating-documents

In [44]:
data_list, news_data = random_sample(newsgroups)
print(len(news_data))
df = pd.DataFrame(news_data, columns = ['text'])
df['text']

2000


0       From: bh437292@longs.LANCE.ColoState.Edu (Basi...
1       Nntp-Posting-Host: surt.ifi.uio.no\nFrom: Thom...
2       From: lieuwen@allegra.att.com (Dan Lieuwen)\nS...
3       From: goldberg@oasys.dt.navy.mil (Mark Goldber...
4       From: johnsd2@rpi.edu (Dan Johnson)\nSubject: ...
                              ...                        
1995    From:  (Phil Bowermaster)\nSubject: C. S. Lewi...
1996    From: randy@megatek.com (Randy Davis)\nSubject...
1997    From: jodfishe@silver.ucs.indiana.edu (joseph ...
1998    From: un026300@wvnvms.wvnet.edu\nSubject: Re: ...
1999    Subject: Re: Jewish Baseball Players?\nFrom: f...
Name: text, Length: 2000, dtype: object

In [45]:
import os
import openai
import tiktoken
from bertopic.representation import OpenAI
from bertopic import BERTopic

from google.colab import userdata
api_key = userdata.get('OpenAI')

# Tokenizer
tokenizer= tiktoken.encoding_for_model("gpt-3.5-turbo")

#Create API model
client = openai.OpenAI(api_key=api_key)
representation_model = OpenAI(
    client,
    model="gpt-3.5-turbo",
    delay_in_seconds=2,
    chat=True,
    nr_docs=4,
    doc_length=100,
    tokenizer=tokenizer
)

# Use the representation model
topic_model = BERTopic(representation_model=representation_model)

In [47]:
def AI_BERTopic_coherence(data,num_topics):

    best_model = topic_model
    best_score = 0
    best_topic_num = 0
    best_topics = []
    best_probs = []

    for i in range(2, num_topics):
        topic_model = BERTopic(representation_model=representation_model, nr_topics = i)
        topics, probs = topic_model.fit_transform(data)

        #Extract tokenized texts
        texts = data.apply(lambda x: x.split()).tolist()

        #Extract the top words per topic from BERTopic
        topics = topic_model.get_topics()
        top_words_per_topic = [
            [word for word, _ in topic_model.get_topics()[topic_id]]
            for topic_id in topic_model.get_topics().keys()
            if topic_id != -1
        ]

        #Create dictionary
        dictionary = corpora.Dictionary(texts)

        #Compute Coherence Score
        coherence_model = CoherenceModel(
            topics=top_words_per_topic,
            texts=texts,
            dictionary=dictionary,
            coherence='c_v'
        )
        coherence_score = coherence_model.get_coherence()

        print(f"Model with {i} topics, coherence score is {coherence_score:}")

        if coherence_score > best_score:
            best_topic_num = i
            best_model = topic_model
            best_score = coherence_score
            best_topics = topics
            best_probs = probs
        else:
            continue

    print(f"Model with {best_topic_num} topics with coherence score {best_score}")
    return best_model, best_score, best_topic_num, best_topics, best_probs

In [46]:
AI_BERTopic_coherence(semantic_df['processed abstract'],10)

RateLimitError: Error code: 429 - {'error': {'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.', 'type': 'insufficient_quota', 'param': None, 'code': 'insufficient_quota'}}

# **Question 4 (35 Points)**


**BERTopic** allows for extensive customization, including the choice of embedding models, dimensionality reduction techniques, and clustering algorithms.

**Dataset Link**: 20 Newsgroup Dataset (Random 2000 values)

4)

4.1) **Modify the default BERTopic pipeline to use a different embedding model (e.g., Sentence-Transformers) and a different clustering algorithm (e.g., DBSCAN instead of HDBSCAN).

4.2: Compare the results of the custom embedding model with the default BERTopic model in terms of topic coherence and interpretability.

4.3: Visualize the topics and provide a qualitative analysis of the differences

**

Usefull Link :https://www.pinecone.io/learn/bertopic/

In [50]:
from sklearn.cluster import DBSCAN
from sentence_transformers import SentenceTransformer

In [48]:
data_list, news_data = random_sample(newsgroups)
print(len(news_data))
df = pd.DataFrame(news_data, columns = ['text'])
df['text']

2000


0       From: jcj@tellabs.com (jcj)\nSubject: Re: When...
1       From: mikeq@freddy.CNA.TEK.COM (Mike Quigley)\...
2       From: dcd@se.houston.geoquest.slb.com (Dan Day...
3       From: nlu@Xenon.Stanford.EDU (Nelson Lu)\nSubj...
4       From: slack@boi.hp.com (David Slack)\nSubject:...
                              ...                        
1995    From: parr@acs.ucalgary.ca (Charles Parr)\nSub...
1996    From: brian@porky.contex.com (Brian Love)\nSub...
1997    From: wdm@world.std.com (Wayne Michael)\nSubje...
1998    From: amehdi@src.honeywell.com (Hossien Amehdi...
1999    From: jeremi@ee.ualberta.ca (William Jeremiah)...
Name: text, Length: 2000, dtype: object

In [76]:
# apply function to all reviews
df['cleaned_text'] = df['text'].apply(lambda x: preprocess(x))

In [77]:
df['cleaned_text']

0       [from, jcj, tellab, com, jcj, subject, peopl, ...
1       [from, mikeq, freddi, cna, tek, com, mike, qui...
2       [from, dcd, houston, geoquest, slb, com, dan, ...
3       [from, nlu, xenon, stanford, edu, nelson, subj...
4       [from, slack, boi, com, david, slack, subject,...
                              ...                        
1995    [from, parr, ac, ucalgari, charl, parr, subjec...
1996    [from, brian, porki, contex, com, brian, love,...
1997    [from, wdm, world, std, com, wayn, michael, su...
1998    [from, amehdi, src, honeywel, com, hossien, am...
1999    [from, jeremi, ualberta, william, jeremiah, su...
Name: cleaned_text, Length: 2000, dtype: object

In [84]:
stop_words = stopwords.words('english')
stop_words.extend(['from', 'subject', 're', 'edu', 'use'])  # custom ones

#Remove stopwords from each token list
df['cleaned_text'] = df['cleaned_text'].apply(
    lambda tokens: [word for word in tokens if word.lower() not in stop_words]
)

In [85]:
#Load a Sentence-Transformers embedding model
embedding_model = SentenceTransformer('all-MiniLM-L6-v2')

#Define a custom clustering model (DBSCAN)
clustering_model = DBSCAN(eps=0.5, min_samples=5, metric='cosine')

#Create BERTopic with custom models
topic_model = BERTopic(
    embedding_model=embedding_model,
    hdbscan_model=clustering_model,  # despite the name, this param accepts any clustering model
    language="english",
    calculate_probabilities=True,  # optional
    verbose=True
)

In [86]:
docs = df['cleaned_text'].apply(lambda tokens: ' '.join(tokens) if isinstance(tokens, list) else tokens)
topics, probs = topic_model.fit_transform(docs)

2025-04-08 04:49:41,833 - BERTopic - Embedding - Transforming documents to embeddings.


Batches:   0%|          | 0/63 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [72]:
texts = df['cleaned_text']

#Extract top words per topic from the BERTopic model
#Skip topic -1 (outliers)
topics = topic_model.get_topics()
top_words_per_topic = [
    [word for word, _ in words]
    for topic_id, words in topics.items()
    if topic_id != -1
]

#Create a Gensim Dictionary
dictionary = corpora.Dictionary(texts)

#Create Coherence Model
coherence_model = CoherenceModel(
    topics=top_words_per_topic,
    texts=texts,
    dictionary=dictionary,
    coherence='c_v'
)

#Get the Coherence Score
coherence_score = coherence_model.get_coherence()
print(f"Coherence Score: {coherence_score:.4f}")

Coherence Score: 0.7086


In [73]:
topic_model.get_topic_info()

Topic  Count                    Name  \
0      0   2000  0_edu_line_subject_com   

                                      Representation  \
0  [edu, line, subject, com, from, organ, write, ...   

                                 Representative_Docs  
0  [from marshal csugrad edu kevin marshal subjec...

In [74]:
topic_model.visualize_topics()

In [75]:
topic_model.visualize_barchart(top_n_topics=12, n_words = 10, width = 350, height = 350)

##Difference in Coherence Scores

We can see that with the customized version of BERTopic with SentenceTransformer model 'all-MiniLM-L6-v2' and clustering model DBSCAN our total coherence score was 0.51, whereas the previous default version of BERTopic
received a coherence score of 0.2547.

## Extra Question (5 Points)

**Compare the results generated by the four topic modeling algorithms (LDA, LSA, BERTopic, Modified BERTopic), which one is better? You should explain the reasons in details.**

**This question will compensate for any points deducted in this exercise. Maximum marks for the exercise is 100 points.**

In [ ]:
# Write your code here

# Mandatory Question

**Important: Reflective Feedback on this exercise**

Please provide your thoughts and feedback on the exercises you completed in this assignment.

Consider the following points in your response:

**Learning Experience:** Describe your overall learning experience in working with text data and extracting features using various topic modeling algorithms. Did you understand these algorithms and did the implementations helped in grasping the nuances of feature extraction from text data.

**Challenges Encountered:** Were there specific difficulties in completing this exercise?

Relevance to Your Field of Study: How does this exercise relate to the field of NLP?

**(Your submission will not be graded if this question is left unanswered)**



In [ ]:
# Your answer here (no code for this question, write down your answer as detail as possible for the above questions):

'''
I thought that the exercises were useful and interesting in trying to group texts
and derive the overall subject/topic similarity between them. It could be useful
for many different types of projects and tasks where information extraction, and
grouping of information is useful. I am still not 100% sure of what all the different
methodologies behind the different types models were doing and what the pros and cons of one versus
the other were, but it was interesting to see how easy they are to implement and
change based on the Gensim, BERTopic libraries, and the extensive amount of documentation
available for each. For instance, I do understand that basically they are calculating
the frequencies of words, and the relative closeness of those words to determine
the liklihood of a given text being a part of one of the topics based on the word count,
and the space between other topic word probabilities and embeddings. I found one
of the major challenges was that trying to run tasks which would change the created
models topic number would take a ton of time to train,
with the BERTopic training model taking approx. 36 minutes to do 10 different topic
amounts. I also found it difficult to connect the model with the API as I would receive
an error about the billing/API credits required.
'''